### PR TITLE
AMP integration with ZEDO adserver

### PR DESCRIPTION
**Pull request by skazedo**

### CODE REVIEW

File: ads/zedo.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/zedo.js]: 2016

File: ads/zedo.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+<!--
+Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/zedo.md]: 2016

File: ads/zedo.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function zedo(global, data) {
+  // check mandatory fields
+  validateData(data, [], [
+    'superId', 'network', 'placementId',
```
Comment (zhouyx) [inline ads/zedo.js]: I see that most of these variables here is required according to your doc. But you are putting them in the optional fields check array? Am I missing something

File: examples/zedo.amp.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<!doctype html>
```
Comment (zhouyx) [inline examples/zedo.amp.html]: is there a special reason that we need a testing page for your ad? This file is not needed I think

**Comment by zhouyx** [comment]
we have a [renderStart API](https://github.com/ampproject/amphtml/blob/master/ads/README.md#methods-available-to-the-ad) that helps AMP to provide better ad loading experience. 
Could you take a look and implement it (renderStart + noContentAvailable)?
Let me know if there's any question.

**Comment by skazedo** [comment]
I am making a call to window.context.renderStart at the time of loading the ad in our client code. Is it required at any other place?

File: ads/zedo.js (hunk 3 above)
Comment (skazedo) [inline ads/zedo.js]: fixed.

File: examples/zedo.amp.html (hunk 4 above)
Comment (skazedo) [inline examples/zedo.amp.html]: removed the file.

**Comment by zhouyx** [comment]
[See comment here](
https://github.com/ampproject/amphtml/issues/5234#issuecomment-250185610)

render-start should be called right after ad is rendered.

File: ads/zedo.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function zedo(global, data) {
+  // check mandatory fields
+  validateData(data, ['superId', 'network', 'placementId',
+  'channel', 'publisher', 'dim', 'renderer'], ['charset', 'callback',
+  'tmy', 'g', 'renderer']);
```
Comment (zhouyx) [inline ads/zedo.js]: `validateData` only validate `data` object. Please put data attribute here. for example 'g' should be 'geo'. 
Also please remove duplicate, if it's mandatory you don't need it to be optional again. 
Also please update all these variables in your readme doc.

**Comment by skazedo** [comment]
Ok. So as i understand, you expect it to be invoked soon after the adserver call but before loading the adcontents. I will have to work on that change and push to production. Will let you know once i am done.

File: ads/zedo.js (hunk 5 above)
Comment (skazedo) [inline ads/zedo.js]: Ok. Fixed.

**Comment by skazedo** [comment]
I signed it!

**Comment by skazedo** [comment]
@googlebot Is there anything else you need from us to complete signing of the CLA?

**Comment by skazedo** [comment]
@zhouyx , I have released the changes for renderStrart and noContentAvailable in our client code. Please let me know if there is anything else that i need to address. Thanks.

**Comment by skazedo** [comment]
@zhouyx , hi, i just figured out that i also need to renderStartImplemented into the config file. Did so and committed but the build failed with "fatal: Invalid symmetric difference expression".  Can you help me. Thanks.

**Comment by zhouyx** [comment]
@skazedo This looks great to me! Thanks for the pull request.
Ping @lannka to take a final look

File: ads/zedo.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ * @param {!Object} data
+ */
+export function zedo(global, data) {
+  // check mandatory fields
+  validateData(data, ['superId', 'network', 'placementId',
+  'channel', 'publisher', 'dim'], ['charset', 'callback', 'renderer']);
+
+  loadScript(global, 'https://ss3.zedo.com/gecko/tag/Gecko.amp.min.js', () => {
+    const ZGTag = global.ZGTag;
+    const charset = data.charset ? data.charset : '';
```
Comment (lannka) [inline ads/zedo.js]: nitpick: `const charset = data.charset || ''`

File: ads/zedo.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    const callback = data.callback ? data.callback : () => {};
```
Comment (lannka) [inline ads/zedo.js]: same here

File: ads/zedo.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function zedo(global, data) {
+  // check mandatory fields
+  validateData(data, ['superId', 'network', 'placementId',
+  'channel', 'publisher', 'dim'], ['charset', 'callback', 'renderer']);
```
Comment (lannka) [inline ads/zedo.js]: for better readability, can you break lines before each '[' and indent 4 spaces on the 2nd line?

File: examples/ads.amp.html [hunk 9]
```diff
@@ -986,6 +987,18 @@ <h2 id="webediads">Webediads</h2>
     <div fallback></div>
   </amp-ad>
 
+  <h2 id="zedo">ZEDO</h2>
+  <amp-ad width="300" height="250"
+          type="zedo"
```
Comment (lannka) [inline examples/ads.amp.html]: please follow the same indentation rule as the other examples.

**Comment by skazedo** [comment]
Thanks @lannka . I have updated with the suggested changes.

**Comment by skazedo** [comment]
@zhouyx anything else i need to address?

**Comment by zhouyx** [comment]
@skazedo sorry for the delay. merged!

**Comment by skazedo** [comment]
thanks @zhouyx

**Pull request merged**